core/embed/rust/src/micropython/map.rs

```diff
@@ -0,0 +1,171 @@
+        unsafe {
+            ffi::mp_map_init(map.as_mut_ptr(), capacity);
+            map.assume_init()
+        }
+    }
+
+    pub fn len(&self) -> usize {
+        self.used()
+    }
+
+    pub fn elems(&self) -> &[MapElem] {
+        // SAFETY: `self.table` should always point to an array of `MapElem` of
+        // `self.len()` items valid at least for the lifetime of `self`.
+        unsafe { slice::from_raw_parts(self.table, self.len()) }
+    }
+
+    pub fn contains_key(&self, index: impl Into<Obj>) -> bool {
+        self.get_obj(index.into()).is_ok()
+    }
+
+    pub fn get(&self, index: impl Into<Obj>) -> Result<Obj, Error> {
+        self.get_obj(index.into())
+    }
+
+    pub fn get_obj(&self, index: Obj) -> Result<Obj, Error> {
+        // SAFETY:
+        //  - `mp_map_lookup` returns either NULL or a pointer to a `mp_map_elem_t`
+        //    value with a lifetime valid for the whole lifetime of the passed immutable
+        //    ref of `map`.
+        //  - with `_mp_map_lookup_kind_t_MP_MAP_LOOKUP`, `map` stays unmodified and the
+        //    cast to mut ptr is therefore safe.
+        unsafe {
+            let map = self as *const Self as *mut Self;
+            let elem = ffi::mp_map_lookup(map, index, ffi::_mp_map_lookup_kind_t_MP_MAP_LOOKUP)
+                .as_ref()
+                .ok_or(Error::Missing)?;
+            Ok(elem.value)
+        }
+    }
+
+    pub fn set(&mut self, index: impl Into<Obj>, value: impl Into<Obj>) {
+        self.set_obj(index.into(), value.into())
+    }
+
+    pub fn set_obj(&mut self, index: Obj, value: Obj) {
+        unsafe {
+            let map = self as *mut Self;
+            let elem = ffi::mp_map_lookup(
+                map,
+                index,
+                ffi::_mp_map_lookup_kind_t_MP_MAP_LOOKUP_ADD_IF_NOT_FOUND,
+            )
+            .as_mut()
+            .unwrap();
+            elem.value = value;
+        }
+    }
+
+    pub fn delete(&mut self, index: impl Into<Obj>) {
+        self.delete_obj(index.into())
+    }
+
+    pub fn delete_obj(&mut self, index: Obj) {
+        unsafe {
+            let map = self as *mut Self;
+            ffi::mp_map_lookup(
+                map,
+                index,
+                ffi::_mp_map_lookup_kind_t_MP_MAP_LOOKUP_REMOVE_IF_FOUND,
+            );
+        }
+    }
+}
+
+impl Clone for Map {
+    fn clone(&self) -> Self {
+        let mut map = Self::with_capacity(self.len());
+        unsafe {
+            ptr::copy_nonoverlapping(self.table, map.table, self.len());
+        }
+        map.set_used(self.used());
+        map.set_all_keys_are_qstrs(self.all_keys_are_qstrs());
+        map.set_is_ordered(self.is_ordered());
+        map.set_is_fixed(0);
+        map
+    }
+}
+
+impl Default for Map {
+    fn default() -> Self {
+        Self::with_capacity(0)
+    }
+}
+
+// SAFETY: We are in a single-threaded environment.
+unsafe impl Sync for Map {}
+unsafe impl Sync for MapElem {}
+
+pub struct MapRef<'a> {
```

Comment:
`Map`, or actually `ffi::mp_map_t`, has a raw pointer field to a table with its elements. `Map` itself does not have any lifetime parameter, so this effectively only models data that is `'static`. `MapRef` is tying a private `Map` value to a lifetime `'a`, exposing it only by-ref within the `'a` lifetime. It's not possible to move the value out from the reference, and because `Map` does not impl `Clone`, it is also impossible to clone it.

All this only holds for immutable ref, therefore we only impl `Deref`, not `DerefMut`, as with a mut ref you could just `core::mem::replace` it out. Better solution would be to make `Map` a wrapper type instead of a type alias and add a lifetime parameter, maybe defaulting to `'static`. But this is simpler for now.

Reply:
Deserves a code comment.